### DISCUSSION TITLE
Blazor -> System.Net.WebClient is not supported on this platform. Use System.Net.Http.HttpClient instead

### DISCUSSION THREAD

**Discussion by rfcdejong** [MarimerLLC/csla#2090] [Questions] [answered]
I'm trying to reuse our business objects inside blazor and now I bump into the following exception:

> System.Net.WebClient is not supported on this platform. Use System.Net.Http.HttpClient instead

breakpoints into csla code aren't being even though symbols are loaded... but we have sync factory methods. And A LOT of them.. 
Could Sync methods use HttpClient? What is the reason WebClient is being used?

Ofcourse I made entry factory method async, but businessrules use properties that trigger 'lazy loading' and also properties calling factory methods tehmself..

Wouldn't it be possible to replace WebClient (at least for blazor) to use HttpClient for Sync calls? 

```
 var response = client.GetAsync("https://myurl/..").Result;
 if (response.IsSuccessStatusCode)
 ....
```

My code in blazor
![image](https://user-images.githubusercontent.com/2665653/106928893-fce77980-6713-11eb-98ac-458a3e3af6ec.png)

comes back with from the server and then during business rules or whatever it throws the follow exception, and I have no clue where in the code it goes wrong cause breakpoints not being hit inside csla
![image](https://user-images.githubusercontent.com/2665653/106928048-11774200-6713-11eb-95e2-b157886e9e16.png)

**Comment by rockfordlhotka** [comment, marked as answer]
The answer is that you can't use sync calls.

I'm unaware of any way to get `HttpClient` to act in a sync manner, at least not without all sorts of deadlock edge cases.

But really, in today's world, why would you ever want to use the sync model anyway? It just isn't the way the world works anymore.

**Reply by rfcdejong** [reply]
I agree, but simply most factory methods are from 2011-2013 and all sync, or at least async with callback event.. team didn't bother to change to async-await

**Reply by rockfordlhotka** [reply]
If you can figure out a reliable way to use `HttpClient` to make sync calls, I suggest you create your own `HttpProxy` type and use that.

I tried to figure it out a long time ago and wasn't able to resolve it. On the other hand, I needed a solution that would work with console/winforms/wpf/uwp/asp/blazor/etc. and you maybe only need to worry about Blazor?

**Reply by rfcdejong** [reply]
I'm going to test some dirty code in a custom proxy with HttpClient that seems like a hack, but it is proven 'non-deadlock' code.
It will be like this code.
Creating a Task.Factory.StartNew to run a function with .ConfigureAwait(false).GetAwaiter.GetResult()

![image](https://user-images.githubusercontent.com/2665653/107013224-5db99500-679a-11eb-9d76-a689ba14e1cb.png)

**Reply by rfcdejong** [reply]
I have to say Blazor is hard to debug.. a breakpoint in the custom Proxy I just created to figure out the calling code... nothing to see in the stacktrace, just a bunch of wasm and mono stuff... and dotnet javascript. So I don't know which code is calling it sync :(

![image](https://user-images.githubusercontent.com/2665653/107018778-27cbdf00-67a1-11eb-8e40-ce2ef24fc362.png)

**Reply by rfcdejong** [reply]
Tried this..
![image](https://user-images.githubusercontent.com/2665653/107019772-5c8c6600-67a2-11eb-9828-0529e18b3972.png)

Sadly a dead end as the mono runtime is single threaded

![image](https://user-images.githubusercontent.com/2665653/107019703-4aaac300-67a2-11eb-88f6-fe585363cd4d.png)

**Reply by rfcdejong** [reply]
At the moment I think my research is ending... we cannot reuse our Csla objects in client side blazor/ WASM / mono. Not because of this topic problem, which I fixed, but this..

![image](https://user-images.githubusercontent.com/2665653/107039114-be0dfe00-67bd-11eb-969b-fb0568a47648.png)

resulting into (somewhere)

![image](https://user-images.githubusercontent.com/2665653/107039225-e0078080-67bd-11eb-91e1-a94143ccde0b.png)

**Reply by rockfordlhotka** [reply]
You should be able to just add async factories to your classes, or directly call the data portal's async methods. In most cases the code in a business class won't care if the data portal is used in sync or async mode (exceptions being lazy loaded properties or rules that invoke the data portal synchronously).

**Reply by rfcdejong** [reply]
I indeed found properties calling sync calls, those being registered as private fields as well as rules invoking sync calls. Also non-async methods in a huge chain of static classes. Kinda dirty code, but the team did it that way. So much code from 2011-2013 and when it went from WcfPortal to HttpPortol things got even dirtier. I just slashed away and used fake data into being able to continue my research if our business objects will be useable in blazor.

The last 'memory access out of bounds' is being caused somewhere at the client-side running CheckBusinessRules. Yes, we have to call that clientside, because of the use of so many private properties, don't ask me why the team did that.

I have to use a quick & dirty Console.WriteLine("...") just to be able poor-mans debugging, because somehow breakpoints aren't being hit in the business libraries, while breakpoints in blazor C# code are being hit. Stepping in sometimes works, but not always. Using git clean -fxd nor rebuilding doesn't change it. Also breaking on all exceptions doesn't do what I want. Just blazor client-side debugging is awful tbh.

If I can't get it to work early next week I have a last shot of trying out blazor server-side as a last resort. If that doesn't work the web-based client will be in VueJS which will need a REST API. This because other teams of the company already use VueJS. It might result in starting with just a database and perhaps the API as a layer using the business.

**Reply by JacoJordaan** [reply]
I bumped in the same issue with Blazor Wasm. Blazor server is working fine with sync calls.

Also have the same issues with Blazor Wasm debugging - breakpoints are not hit inside csla library.

For Blazor Wasm I am still busy converting all sync calls to Async. 

The easiest is the factory methods. 

The properties are also easy except if you are using those properties in business rules since the business rule may expect the property to already be loaded, but it may still be busy with Async call and have its default value.
For those properties I opted to use unit of work class to load properties at the time of loading the object.

All sync business rules calling a non-local dataportal method must be converted to an async rule.

**Reply by rfcdejong** [reply]
I found the failing rule, for "memory access out of bounds". I don't know why it fails, maybe because of IComparable or params.
This rule is only being used in one place, though it seems like a generic rule.

            BusinessRules.AddRule(new EqualsValueSetBerekening(IsJaarlijksHerberekenenMogelijkProperty, BudgetpostFilterProperty,
                AandBudgetpostFilter.HypotheekWoonEnVasteLasten, AandBudgetpostFilter.HypotheekWoonVasteEnOverigeLasten));

...
I had to debug it with Console.WriteLine, it fails at:
result = _values.Any(value.Equals);


    public sealed class EqualsValueSetBerekening : UNIT4CommonBusinessRule
    {
        private readonly IPropertyInfo _valueProperty;
        private readonly IComparable[] _values;

        public EqualsValueSetBerekening(IPropertyInfo primaryProperty, IPropertyInfo valueProperty, params IComparable[] values)
            : base(primaryProperty)
        {
            InputProperties = new List<IPropertyInfo> { valueProperty };
            _valueProperty = valueProperty;
            _values = values;

            int valueCounter = 1;
            foreach (IComparable value in values)
            {
                RuleUri.AddQueryParameter(String.Format("value{0}", valueCounter++), value.ToString());
            }
        }

        protected override void ExecuteRule(IRuleContext context)
        {
            Console.WriteLine("EqualsValueSetBerekening - begin");
            var value = context.InputPropertyValues[_valueProperty] as IComparable;
            if (value == null)
            {
                context.AddOutValue(false);
                return;
            }
            Console.WriteLine("EqualsValueSetBerekening - reading values");
            var result = _values.Any(value.Equals);
            context.AddOutValue(result);
            Console.WriteLine("EqualsValueSetBerekening - end");
        }
    }

I made a new rule to replace it, and the memory error is resolved. weird stuff!

        public sealed class IsJaarlijksHerberekenenMogelijkRule : UNIT4CommonBusinessRule
        {
            private readonly PropertyInfo<AandBudgetpostFilter> _budgetpostFilterProperty;

            public IsJaarlijksHerberekenenMogelijkRule(PropertyInfo<bool> isJaarlijksHerberekenenMogelijkProperty, PropertyInfo<AandBudgetpostFilter> budgetpostFilterProperty) 
                : base(isJaarlijksHerberekenenMogelijkProperty)
            {
                _budgetpostFilterProperty = budgetpostFilterProperty;
                InputProperties = new List<IPropertyInfo> { budgetpostFilterProperty };
            }

            protected override void ExecuteRule(IRuleContext context)
            {
                var budgetpostFilter = (AandBudgetpostFilter)context.InputPropertyValues[_budgetpostFilterProperty];

                switch (budgetpostFilter)
                {
                    case AandBudgetpostFilter.HypotheekWoonEnVasteLasten:
                    case AandBudgetpostFilter.HypotheekWoonVasteEnOverigeLasten:
                        context.AddOutValue(true);
                        break;
                    default:
                        context.AddOutValue(false);
                        break;
                }
            }
        }

**Reply by rockfordlhotka** [reply]
I'm glad you found the issue and a workaround!!
